### PR TITLE
Add rule S6245: Disabling server-side encryption of S3 buckets is security-sensitive

### LINKED ISSUE #3201: Rule S6245: Disabling server-side encryption of S3 buckets is security-sensitive
## WHAT
The rule detects when an S3 bucket is not configured with encryption using the JavaScript AWS CDK  library. It does not cover custom policy statement definitions. This has been previously implemented for Python, see [SONARPY-1014](https://github.com/SonarSource/sonar-python/pull/1128).

Specification ticket: [APPSEC-41](https://jira.sonarsource.com/browse/APPSEC-41)
[MMF-2812](https://jira.sonarsource.com/browse/MMF-2812)

## HOW

Detection logic
- Method: `aws-cdk-lib.aws_s3.Bucket` constructor
- Argument: `props` 3rd positional argument
- Property: `encryption`
  - Safe values: `BucketEncryption.KMS_MANAGED`, `BucketEncryption.S3_MANAGED`, `BucketEncryption.KMS`
  - Sensitive value: `BucketEncryption.UNENCRYPTED`
  - Default value: when not set `encryption` value defaults to `BucketEncryption.UNENCRYPTED`

The rule raises on the _Bucket_ construct class when any of the above conditions are met.
```javascript
const bucketNoncompliant1 = new s3.Bucket(this, 'id1', { 
            bucketName: 'bucketnoncompliant1'
}); // Sensitive: Omitting 'encryption' implies no server-side encryption. Make sure it is safe here

------
new s3.Bucket(this, 'id2', {
            bucketName: 'bucketnoncompliant2',
            encryption: s3.BucketEncryption.UNENCRYPTED, // Sensitive: Objects in the bucket are not encrypted. Make sure it is safe here
});

-----
const encryption = s3.BucketEncryption.UNENCRYPTED; // Sensitive: Propagated setting

const bucketNoncompliant3 = new s3.Bucket(this, 'id3', {
           bucketName: 'bucketnoncompliant3',
           encryption: encryption, // Sensitive: Objects in the bucket are not encrypted. Make sure it is safe here
           removalPolicy: cdk.RemovalPolicy.DESTROY
});
```

## Code examples

https://github.com/SonarSource/security-expected-issues/pull/722

## Message and highlighting

See https://github.com/SonarSource/rspec/pull/1088

### PR DESCRIPTION
**Pull request by yassin-kammoun-sonarsource**
Fixes #3201

### CODE REVIEW

**Comment by yassin-kammoun-sonarsource** [comment]
Depends on #3298

File: eslint-bridge/tests/rules/comment-based/aws-s3-bucket-server-encryption.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+new s3.Bucket(this, 'id5', {
+  bucketName: 'bucketCompliant2',
+  encryption: this.BucketEncryption.UNENCRYPTED, // Compliant
+});
+
+const s3Fake = require('some-fake-aws-cdk-lib/aws-s3');
+new s3.Bucket(this, 'id6', {
+  bucketName: 'bucketCompliant3',
+  encryption: s3Fake.BucketEncryption.UNENCRYPTED, // Compliant
```
Comment (ilia-kebets-sonarsource) [inline eslint-bridge/tests/rules/comment-based/aws-s3-bucket-server-encryption.js:35]: Is this not noncompliant?

File: eslint-bridge/tests/rules/comment-based/aws-s3-bucket-server-encryption.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+});
+
+new s3.Bucket(this, 'id4', {
+  bucketName: 'bucketCompliant1',
+  encryption: s3.BucketEncryption.KMS_MANAGED // Compliant
+});
+
+new s3.Bucket(this, 'id5', {
+  bucketName: 'bucketCompliant2',
+  encryption: this.BucketEncryption.UNENCRYPTED, // Compliant
```
Comment (ilia-kebets-sonarsource) [inline eslint-bridge/tests/rules/comment-based/aws-s3-bucket-server-encryption.js:29]: Is this not noncompliant?

File: eslint-bridge/src/utils/utils-module.ts [hunk 3]
```diff
@@ -196,3 +197,37 @@ export function getModuleAndCalledMethod(callee: estree.Node, context: Rule.Rule
   }
   return { module, method };
 }
+
+/**
+ * Checks that an ESLint member expression matches a fully qualified name
```
Comment (ilia-kebets-sonarsource) [inline eslint-bridge/src/utils/utils-module.ts:202]: Could you add an example?

File: eslint-bridge/tests/rules/comment-based/aws-s3-bucket-server-encryption.js (hunk 2 above)
Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/tests/rules/comment-based/aws-s3-bucket-server-encryption.js:29]: `this` keyword in `this.BucketEncryption.UNENCRYPTED`

File: eslint-bridge/src/utils/utils-module.ts (hunk 3 above)
Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/utils/utils-module.ts:202]: 👍

File: eslint-bridge/src/rules/aws-s3-bucket-server-encryption.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+      });
+    }
+
+    function isUnencrypted(encrypted: MemberExpression) {
+      return hasFullyQualifiedName(
+        context,
+        encrypted,
+        'aws-cdk-lib/aws-s3',
+        'BucketEncryption',
+        'UNENCRYPTED',
```
Comment (ilia-kebets-sonarsource) [inline eslint-bridge/src/rules/aws-s3-bucket-server-encryption.ts:74]: so we only handle the explicitly unencrypted case?

File: eslint-bridge/src/rules/aws-s3-bucket-server-encryption.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    const encryptedValue = getValueOfExpression(
+      context,
+      encryptedProperty.value,
+      'MemberExpression',
+    );
+    if (encryptedValue && isUnencrypted(encryptedValue)) {
+      const secondary = { locations: [] as Node[], messages: [] as string[] };
+      const propagatedProperty = encryptedProperty.value !== encryptedValue;
```
Comment (ilia-kebets-sonarsource) [inline eslint-bridge/src/rules/aws-s3-bucket-server-encryption.ts:57]: `propagatedProperty` => `isPropagatedProperty`

File: eslint-bridge/src/rules/aws-s3-bucket-server-encryption.ts (hunk 4 above)
Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/rules/aws-s3-bucket-server-encryption.ts:74]: Yes, that's basically the same behaviour as [sonar-python's implementation](https://github.com/SonarSource/sonar-python/blob/a92edccf9b17945353ace9af3b47016f0f942f2c/python-checks/src/main/java/org/sonar/python/checks/cdk/S3BucketServerEncryptionCheck.java#L57).

File: eslint-bridge/src/rules/aws-s3-bucket-server-encryption.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      if (propagatedProperty) {
+        secondary.locations = [(encryptedValue as any).parent as Node];
+        secondary.messages = ['Propagated setting.'];
```
Comment (ilia-kebets-sonarsource) [inline eslint-bridge/src/rules/aws-s3-bucket-server-encryption.ts:60]: add the value to messages

File: eslint-bridge/src/rules/aws-s3-bucket-server-encryption.ts (hunk 5 above)
Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/rules/aws-s3-bucket-server-encryption.ts:57]: 👍

File: eslint-bridge/src/rules/aws-s3-bucket-server-encryption.ts (hunk 6 above)
Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/rules/aws-s3-bucket-server-encryption.ts:60]: 👍

**Review by ilia-kebets-sonarsource — APPROVED** [review]
Just the extra tab in comment-based test file, extract the parent and the example, otherwise LGTM

**Pull request merged**